rts/lra/lra-service-base/src/main/java/io/narayana/lra/RequestBuilder.java

```diff
@@ -0,0 +1,163 @@
+import org.apache.http.entity.ContentType;
+import org.apache.http.entity.StringEntity;
+import org.apache.http.protocol.HTTP;
+
+import javax.ws.rs.WebApplicationException;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+public class RequestBuilder {
+    private HttpRequestBase httpRequest;
+    private ArrayList<String> paths;
+
+    private URIBuilder builder;
+    private boolean async;
+    private long timeout;
+    private TimeUnit timeUnit;
+
+    public RequestBuilder(URI baseURI) {
+        builder = new URIBuilder(baseURI);
+        paths = new ArrayList<>();
+
+        Collections.addAll(paths, strip(strip(baseURI.getPath())).split("/"));
+    }
+
+    private String strip(String path) {
+        return path.replaceAll("^/+", "").replaceAll("/+$", "");
+    }
+
+    public RequestBuilder path(String path) {
+        Collections.addAll(paths, strip(path).split("/"));
+
+        return this;
+    }
+
+    public RequestBuilder queryParam(String name, long value) {
+        builder.setParameter(name, Long.toString(value));
+        return this;
+    }
+
+    public RequestBuilder queryParam(String name, String value) {
+        builder.setParameter(name, value);
+        return this;
+    }
+
+    public RequestBuilder queryParam(String name, boolean value) {
+        builder.setParameter(name, Boolean.toString(value));
+        return this;
+    }
+
+    /**
+     * create a request. This method must be called before adding
+     * request headers
+     * @return the builder
+     */
+    public RequestBuilder request() {
+        builder.setPathSegments(paths);
+
+        httpRequest = new HttpGet();
+        return this;
+    }
+
+    /**
+     * add a request header. Must be called after calling {@link RequestBuilder#request()}
+     * @param name the name of the header
+     * @param value the header value
+     * @return the builder
+     */
+    public RequestBuilder header(String name, String value) {
+        httpRequest.addHeader(name, value);
+        return this;
+    }
+
+    /**
+     * add a request header. Must be called after calling {@link RequestBuilder#request()}
+     * @param name the name of the header
+     * @param value the header value
+     * @return the builder
+     */
+    public RequestBuilder header(String name, URI value) {
+        httpRequest.addHeader(name, value == null ? null : value.toASCIIString());
+        return this;
+    }
+
+    public RequestBuilder async(long timeout, TimeUnit timeUnit) {
+        this.async = true;
+        this.timeout = timeout;
+        this.timeUnit = timeUnit;
+        return this;
+    }
+
+    private HttpRequestBase setContent(HttpEntityEnclosingRequestBase req, String content, String mediaType) {
```

Comment:
mediaType is ignored. This seems to be important.

Reply:
We always call it with MediaType.TEXT_PLAIN. But I want to keep the API calls as close as possible to what we had before. If you like I can check for an unsupported media type by replacing it with:

```java
    private HttpRequestBase setContent(HttpEntityEnclosingRequestBase req, String content, String mediaType) {
        if (content != null) {
            if (!mediaType.equals(MediaType.TEXT_PLAIN)) {
                throw new IllegalArgumentException("RequestBuilder: only content type mediaType is supported");
            }
            req.setEntity(new StringEntity(content, HTTP.DEF_CONTENT_CHARSET));
        }

        return req;
    }
```

Reply:
What API are you referring to, please? I think the cleanest would be to remove the parameter if it's not being used as it may lead to false assumptions if someone won't read the body of the method in the future. But it's not that critical so if you want to keep it this way we can refactor later.

Reply:
There are multiple HTTP client implementations:

-  java.net.http.HttpClient (only available in JDK 11)
- resteasy client API
- apache HttpClient (which resteasy uses under the covers)

Let's keep the interface to the HTTP client as generic as possible until we make the move to JDK 11.